core/src/main/scala/org/apache/spark/MapOutputTracker.scala

```diff
@@ -284,6 +291,41 @@ private[spark] class MapOutputTrackerMaster(conf: SparkConf)
     cachedSerializedStatuses.contains(shuffleId) || mapStatuses.contains(shuffleId)
   }
 
+  /**
+   * Return a list of locations which have the largest map outputs given a shuffleId
+   * and a reducerId.
+   *
+   * This method is not thread-safe
+   */
+  def getLocationsWithLargestOutputs(
+      shuffleId: Int,
+      reducerId: Int,
+      numReducers: Int,
+      numTopLocs: Int)
+    : Option[Array[BlockManagerId]] = {
+    if (!shuffleIdToReduceLocations.contains(shuffleId) && mapStatuses.contains(shuffleId)) {
+      // Pre-compute the top locations for each reducer and cache it
+      val statuses = mapStatuses(shuffleId)
+      if (statuses.nonEmpty) {
+        val ordering = Ordering.by[(BlockManagerId, Long), Long](_._2).reverse
+        shuffleIdToReduceLocations(shuffleId) = new HashMap[Int, Array[BlockManagerId]]
+        var r = 0
+        while (r < numReducers) {
+          // Add up sizes of all blocks at the same location
+          val locs = statuses.map { s =>
```

Comment:
Actually, I think we should just go with a straightforward imperative code here.  I don't think it will be significantly longer and can't be slow by accident (it'll also create less garbage).